### PR TITLE
[Broker] Fix NonDurableCursorImpl initialPosition by startCursorPosition greater than lastConfirmedEntry problem.

### PR DESCRIPTION
**Pull request by congbobo184**
## Motivation
In order to fix reader set ```startMessageId``` greater than ```lastConfirmedEntry``` probelm.

When the reader set ```startMessageId``` greater than ```lastConfirmedEntry```, generate ```NonDurableCursorImpl``` will carry a ```MarkdeletePosition``` greater than lastConfirmedEntry.
When use ```getNumberOfEntriesInBacklog``` will throw 
image

## implement
When ```startMessageId``` greater than ```lastConfirmedEntry``` use ```lastConfirmedEntry``` as ```markdeletePosition```
### Verifying this change
Add the tests for it

Does this pull request potentially affect one of the following parts:
If yes was chosen, please highlight the changes

Dependencies (does it add or upgrade a dependency): (no)
The public API: (no)
The schema: (no)
The default values of configurations: (no)
The wire protocol: (no)
The rest endpoints: (no)
The admin cli options: (no)
Anything that affects deployment: (no)

### CODE REVIEW

**Review by eolivelli — COMMENTED** [review]
+1  looks good to me.

But I don't know this part of Pulsar very well.
More eyes are needed

**Review by codelipenghui — APPROVED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorImpl.java [hunk 1]
```diff
@@ -40,7 +40,8 @@
         // Compare with "latest" position marker by using only the ledger id. Since the C++ client is using 48bits to
         // store the entryId, it's not able to pass a Long.max() as entryId. In this case there's no point to require
         // both ledgerId and entryId to be Long.max()
-        if (startCursorPosition == null || startCursorPosition.getLedgerId() == PositionImpl.latest.getLedgerId()) {
+        if (startCursorPosition == null || startCursorPosition.getLedgerId() == PositionImpl.latest.getLedgerId()
+                || startCursorPosition.compareTo(ledger.lastConfirmedEntry) > 0) {
```
Comment (gaoran10) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorImpl.java:44]: It seems that the condition `startCursorPosition.compareTo(ledger.lastConfirmedEntry) > 0` already cover the condition `startCursorPosition.getLedgerId() == PositionImpl.latest.getLedgerId()`, so compare with `ledger.lastConfirmedEntry` is enough?

Comment (congbobo184) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorImpl.java:44]: good suggestion.

Comment (klwilson227) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorImpl.java]: Do we have test case coverage?

**Review by zymap — APPROVED** [review]

**Pull request merged**